services/QuillLMS/lib/tasks/diagnostic_question_optimal_concepts.rake

```diff
@@ -0,0 +1,156 @@
+      puts '  rake diagnostic_question_optimal_concepts:populate_from_csv < path/to/local/file.csv'
+      puts ''
+      puts 'If you are piping data into Heroku, you need to include the --no-tty flag:'
+      puts '  heroku run rake diagnostic_question_optimal_concepts:populate_from_csv -a empirical-grammar --no-tty < path/to/local/file.csv'
+      exit 1
+    end
+
+    diagnostics = DIAGNOSTIC_ID_LOOKUP.transform_values {|id| Activity.find(id)}
+
+    total_count = 0
+    duplicate_count = 0
+    missing_concept_count = 0
+    CSV.parse(pipe_data, headers: true).each do |row|
+      activity = diagnostics[row[DIAGNOSTIC_INDEX]]
+      activity_questions = activity.data['questions'].filter {|d| d['questionType'] != 'titleCards' }
+      question_number = row[QUESTION_NUMBER_INDEX].to_i
+      question_uid = activity_questions[question_number - 1]['key']
+      question = Question.find_by(uid: question_uid)
+      concepts_row = row[OPTIMAL_CONCEPTS_INDEX]
+
+      full_concept_names = concepts_row.split(DELIMITER).reject{|concept| concept == ""}.uniq
+      concepts = full_concept_names.map do |concept_name|
+        level0, level1, level2 = concept_name.split("|").map(&:strip).reverse
+        concepts = Concept.left_outer_joins(:parent)
+          .joins("LEFT OUTER JOIN concepts AS grandparent ON parent.parent_id = grandparent.id")
+          .where(visible: true)
+          .where.not(parent: {parent_id: nil})
+        concepts = concepts.where(parent: {name: level1}) if level1
+        concepts = concepts.where(grandparent: {name: level2}) if level2
+        concept = concepts.where(name: level0)
+        if concept.length > 1
+          puts "Matched too many Concepts: #{concept_name}"
+        elsif concept.length == 1
+          # A number of questions are used in multiple diagnostics, but should have the same Optimal Concepts for each question.  Since each question has a row in the CSV for every diagnostic it's in, we expect to have about 50 duplicates which we can skip over
+          begin
+            DiagnosticQuestionOptimalConcept.create!(concept_id: concept.first.id, question_uid: question.uid)
+          rescue ActiveRecord::RecordNotUnique
+            duplicate_count += 1
+          end
+        else
+          puts "Failed to find Concept: #{concept_name}"
+          missing_concept_count += 1
+        end
+        total_count += 1
+      end
+    end
+    puts "Total rows in CSV processed: #{total_count}"
+    puts "Duplicate question + concept data rows: #{duplicate_count}"
+    puts "Records with failures to match: #{missing_concept_count}"
+  end
+
+  desc "Validate the input CSV to ensure that all data that we expect to find in the LMS using the CSV is found.  This will flag cases where there's some kind of mis-match so that we can manually confirm that any discrepancies are just typos, but that the correct data is being found."
+  task validate_csv_data: :environment do
```

Comment:
Is there any reason not to call the validate task within the main task?

Reply:
Currently the validation doesn't return a "clean" response when doing an automated comparison.  It flags a number of questions as "non-identical" between the CSV and the Notion doc that, when examined manually, are revealed to be minor spacing or wording issues.  We could insist that Curriculum update their Notion doc, but I figured it'd be better to just "sign off" on the discrepancies and move on.  But this means that there's a human step that has to be done when analyzing the validation results.

Reply:
Hmm, I have to agree that it is a little odd that the validation/dry_run step uses entirely different code than the live step. How are we sure it's validating the live code?